sequencer/src/block/full_payload/ns_table.rs

```diff
@@ -221,46 +223,24 @@ impl NsTable {
     ///    must be nonzero.
     /// 3. Header consistent with byte length (obsolete after
     ///    <https://github.com/EspressoSystems/espresso-sequencer/issues/1604>)
-    pub fn validate(&self) -> Result<(), NsTableValidationError> {
+    /// 4. Final offset must equal `payload_byte_len` (obsolete after
+    ///    <https://github.com/EspressoSystems/espresso-sequencer/issues/1604>)
+    pub fn validate(
+        &self,
+        payload_byte_len: &PayloadByteLen,
+    ) -> Result<(), NsTableValidationError> {
         use NsTableValidationError::*;
 
-        // Byte length for a table with `x` entries must be exactly
-        // `x * NsTableBuilder::entry_byte_len() + NsTableBuilder::header_byte_len()`
-        if self.bytes.len() < NsTableBuilder::header_byte_len()
-            || (self.bytes.len() - NsTableBuilder::header_byte_len())
-                % NsTableBuilder::entry_byte_len()
-                != 0
-        {
-            return Err(InvalidByteLen);
-        }
-
-        // Header must declare the correct number of namespaces
-        //
-        // TODO this check obsolete after
-        // https://github.com/EspressoSystems/espresso-sequencer/issues/1604
-        if self.len().0 != self.read_num_nss() {
-            return Err(InvalidHeader);
-        }
+        // conditions 1-3
+        self.validate_deserialization_invariants()?;
 
-        // Namespace IDs and offsets must increase monotonically. Offsets must
-        // be nonzero.
-        {
-            let (mut prev_ns_id, mut prev_offset) = (None, 0);
-            for (ns_id, offset) in self.iter().map(|i| {
-                (
-                    self.read_ns_id_unchecked(&i),
-                    self.read_ns_offset_unchecked(&i),
-                )
-            }) {
-                if let Some(prev_ns_id) = prev_ns_id {
-                    if ns_id <= prev_ns_id {
-                        return Err(NonIncreasingEntries);
-                    }
-                }
-                if offset <= prev_offset {
-                    return Err(NonIncreasingEntries);
-                }
-                (prev_ns_id, prev_offset) = (Some(ns_id), offset);
+        // condition 4
+        let len = self.len().0;
+        if len > 0 {
```

Comment:
What if `len == 0`? Do we want to check that `payload_byte_len == 0` as well?

Reply:
`len` here refers to the number of entries in the namespace table. If it's zero then there is no entry whose offset we must check against `payload_byte_len`. (We need to compute `len-1` just to check anything.)

Perhaps `len` is a confusing name. Maybe `num_nss` or something?

Reply:
Yes that's a good idea.